### PR TITLE
Fixing issue #14850 - Adding null check before calling _asyncStreamOp…

### PR DESCRIPTION
**Pull request by ofirmakmal**
Fixes issue:
https://github.com/dotnet/corefx/issues/14850

### CODE REVIEW

File: src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs [hunk 1]
```diff
@@ -150,8 +150,14 @@ internal bool CancelStreamOperation()
...
                 return false;
 
-            _asyncStreamOperation.Cancel();
-            return true;
+			try
+            {
+				if (_asyncStreamOperation != null)
+					_asyncStreamOperation.Cancel();
+            }
+            catch { }
```
Comment (stephentoub) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: Why is this try/catch needed?

File: src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs [hunk 2]
```diff
@@ -150,8 +150,14 @@ internal bool CancelStreamOperation()
             if (_callbackInvoked)
                 return false;
 
-            _asyncStreamOperation.Cancel();
-            return true;
+			try
+            {
+				if (_asyncStreamOperation != null)
+					_asyncStreamOperation.Cancel();
```
Comment (stephentoub) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: This can just be ```_asyncStreamOperation?.Cancel();```

File: src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs [hunk 3]
```diff
@@ -150,8 +150,14 @@ internal bool CancelStreamOperation()
             if (_callbackInvoked)
                 return false;
 
-            _asyncStreamOperation.Cancel();
-            return true;
+			try
+            {
```
Comment (stephentoub) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: Nit: the spacing is off in the code... looks like a mix of tabs/spaces were used. Tabs shouldn't be used.

File: src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs [hunk 4]
```diff
@@ -150,7 +150,12 @@ internal bool CancelStreamOperation()
             if (_callbackInvoked)
                 return false;
 
-            _asyncStreamOperation.Cancel();
+            try
+            {
+                _asyncStreamOperation?.Cancel();
+            }
+            catch { }
```
Comment (stephentoub) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: Why did you add a try/catch?

File: src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs [hunk 5]
```diff
@@ -150,7 +150,12 @@ internal bool CancelStreamOperation()
             if (_callbackInvoked)
                 return false;
 
-            _asyncStreamOperation.Cancel();
+            try
+            {
+                _asyncStreamOperation?.Cancel();
```
Comment (danmosemsft) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: Perhaps this should set it to null. It's only called in the finalizer.

Comment (ofirmakmal) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: This method is being called from the finalizer.
It's members are not consider reachable either way as soon as the finalizer is done, and will be reclaimed (if no other objects reference them) on the next GC round in that generation.

File: src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: But why is it safe to eat that exception?  In general, we avoid eating all exceptions, an in cases like this typically prefer to just let the process crash.  What exceptions have you seen emerge from this call that you're trying to handle?

File: src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs (hunk 5 above)
Comment (stephentoub) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: > It's members are not consider reachable either way as soon as the finalizer is done

Unless it's resurrected.  I'm not sure if that's what @danmosemsft is concerned about.

Comment (danmosemsft) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: I meant: defensively in case future change adds a call to this, since clearly it's not expected to be called twice. No big deal though

File: src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs (hunk 4 above)
Comment (ofirmakmal) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: NullReferenceException on that field.
Would you like me to remove the try-catch? I've only added it as the close method also do this, and it's been called from the finalizer and it will crash the process.
IMO in this case, eating the exception will no do harm as those classed are about to be reclaimed, and there is no other thing for us to do in case of an exception.

Would you like me to remove the try-catch block?

File: src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs (hunk 5 above)
Comment (ofirmakmal) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: I have added a null set on that field.
Thanks!

File: src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: > NullReferenceException on that field.

Now that you've added the null check, there won't be a null ref from accessing that field.

> Would you like me to remove the try-catch block?

Yes, please.  Until there's good reason to do so, let's let any unexpected exceptions propagate.

Comment (ofirmakmal) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: Done. Thanks for your comments.

**Comment by ofirmakmal** [comment]
Is there anything else I should do to have this PR merged?

File: src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs [hunk 6]
```diff
@@ -150,7 +150,8 @@ internal bool CancelStreamOperation()
             if (_callbackInvoked)
                 return false;
 
-            _asyncStreamOperation.Cancel();
+            _asyncStreamOperation?.Cancel();
+            _asyncStreamOperation = null;
```
Comment (stephentoub) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: Dan's feedback and my feedback conflicted in the resulting code.  There's no value in setting _asyncStreamOperation to null if it's already null.  If we weren't doing that, then it makes sense to do:
```C#
_asyncStreamOperation?.Cancel();
```
instead of the exact equivalent:
```C#
if (_asyncStreamOperation != null)
{
    _asyncStreamOperation.Cancel();
}
```
but since you're also setting the field to null, it makes sense to do:
```C#
if (_asyncStreamOperation != null)
{
    _asyncStreamOperation.Cancel();
    _asyncStreamOperation = null;
}
```
instead of what's currently there, as then we're using the existing null check to avoid setting the field to null when it already is.

Comment (ofirmakmal) [inline src/System.Runtime.WindowsRuntime/src/System/IO/StreamOperationAsyncResult.cs]: Fixed and pushed.
Thanks Stephen.

**Comment by stephentoub** [comment]
Thanks, @ofirmakmal.
